torch/csrc/api/include/torch/data/samplers/base.h

```diff
@@ -37,6 +38,51 @@ class Sampler {
   virtual void load(serialize::InputArchive& archive) = 0;
 };
 
+/// Wraps a provided sampler to make it thread safe.
+template <typename OriginalSampler>
+class ThreadSafeSampler : public Sampler {
```

Comment:
Can we inherit from OriginalSampler directly? Then we don't need to save another copy of sampler_ here.

Reply:
Then we need to create a threadsafe sampler for each different sampler that the user creates. This simply allow us to take in any sampler and use it inside the chunkdataset. This is only needed for the chunk sampler, so we can simply wrap any existing sampler using this approach.